### ISSUE TITLE
Explosion falloff not related to radius

### ISSUE THREAD

**Issue by squeek502** [fortressforever/fortressforever#3]
See: 
- https://github.com/fortressforever/fortressforever/blob/f663579932d1a685a2cb9195d3c25c3393c29b08/game_shared/ff/ff_gamerules.cpp#L1286
- https://github.com/fortressforever/fortressforever/blob/f663579932d1a685a2cb9195d3c25c3393c29b08/game_shared/ff/ff_gamerules.cpp#L1418

This makes falloff unreliable, and makes things that have an altered damage:radius ratio (the IC and frag grens most notably) either do zero damage before getting to the edge of its radius (in the case of the IC) or do more than 0.5x damage at the edge of its radius (in the case of frag grens).

Note: There would be knock-on effects to gameplay if this was fixed, and should only be fixed if we have a plan in place to deal with those.

Here's a commit that reverted some accidentally committed test code that implements a fix and prints the differences: 24eb0aa5e59ea64ee7b52564113551cbd34c4328

**Comment by squeek502** [comment]
I misremembered the effects of this. Turns out most explosions do less than they would if the falloff was "fixed" (meaning 0.5x damage at the edge of the explosion for all explosions).

Damages close to the edge of the explosion radius:

| Thing exploding | Base damage | "Fixed" damage | Current damage |
| --- | --- | --- | --- |
| Frag Grenade | 145.000000 | 68.820503 | 3.148529 |
| IC Rocket | 55.000000 | 26.769991 | -9.159111 |
| RPG Rocket | 102.000000 | 49.961639 | 50.981998 |
| Grellow Pipe | 108.000000 | 50.931957 | 39.306984 |
| MIRV | 180.000000 | 74.475624 | 21.713440 |
| MIRVlet | 145.000000 | 57.599770 | -17.745255 |
| Unbounced fully charged rail | 40.000000 | 20.641932 | -8.395172 |
| Backpack by an EMP | 50.000000 | 33.925419 | -27.157990 |

So... yeah. It has wide-reaching and substantial effects (except on RPG rockets).

**Comment by mikeparker** [comment]
Good findings, nice table.

I think personally, that I agree with valve's initial direction - that if you get hit by an explosion, you should know that you've taken at least a reasonable amount of damage. I think the game is more consistent and easier to read if you don't get hit for 1hp by a mirvlet, or 0hp from an IC rocket!

I think there are two variables we need to decide on:
- Should the radius % be adjustable per explosion, or automatic? TFC's explosions had an automatic radius based on where the damage reached 50% power.
- Should the falloff rate be adjustable per explosion, or automatic? TFC's explosions all had the same falloff rate.

So, the combination of these 2 variables gives us 4 possible outcomes:
1. TFC explosions - fixed falloff, fixed radius
2. Current FF explosions - fixed falloff, adjustable radius
3. Option3: Adjustable falloff, fixed radius (the edge will always be 50% but you can adjust the falloff so explosions are more localised or more spread out)
4. Option4: Adjustable falloff, adjustable radius (the edge will be whatever % we want, and the fall rate can be as fast or slow as we want)

Right now, we're using a hard coded value for the radius rather than a percentage based radius. If we're using a hard coded value it doesn't make sense to use a fixed falloff rate, so the first thing we should change is either varying the falloff rate by using the radius value, or changing the radius value to be percentage based.

I would either vote for option 1 or option 3. I think option 4 is too flexible to sensibly balance and I think option 2 is pointless.

**Comment by squeek502** [comment]
I'm not sure I follow what you mean by fixed/adjustable falloff. Right now, we have a fixed falloff and an adjustable radius. However, that fixed falloff does not account for the adjustable radius, so instead we have a weird hybrid of a fixed falloff that assumes a fixed/damage-based radius while we actually have an adjustable/damage-independent radius.

Given that we're choosing to fix the above inconsistency, I lean towards option 2, as it seems the most straightforward: explosion radius can be adjusted independent of the damage and the falloff will remain consistent.  Option 2 is also what the "Fixed" column shows in the above table.

Option 1 is not an option if we want to keep frag gren damage the same (wouldn't be able to lower the radius of frag grens without decreasing their damage).

I don't understand why option 3 would be desirable; it's simply a _slightly_ more flexible option 1, in a way that won't make much real difference.

And option 4 is just a more flexible option 2, which might end up being useful if the consistent 50% falloff creates too many balance problems.

**Comment by mikeparker** [comment]
To clarify, when i say fixed falloff rate, i mean that no matter what the radius is, a certain distance away from the grenade will always cause the same damage (provided you're inside the radius), which I believe is the current situation. You can "fix" the current situation by either snapping the radius to where 0.5x is (and ignoring all the radius numbers), or "stretching" the damage so that the current radius values are where the 0.5x damage is.

Maybe i should draw some graphs, haha

**Comment by squeek502** [comment]
Okay, I get what you're saying. I'd be really curious to see what the current 50% damage radiuses end up being for things like the frag gren, IC, railgun, etc. For the frag gren, it would cut the radius basically in half by the looks of it.

**Comment by mikeparker** [comment]
Yeah, it means aimed grenades are still just as vicious but unaimed / spammy / dodged grenades could be ignored rather than dealing a medium to small amount of damage.

**Comment by mikeparker** [comment]
Would be interesting to know the blue pipe values too. I guess thats one of the biggest contenders to have its radius reduced, if indeed it's radius is larger than TFC.

**Comment by mikeparker** [comment]
As a simple fix, I wouldnt mind having a lower cutoff at like 25% or something, that would be quite easy to implement and woudlnt really change the gameplay drastically. That would basically mean that if an explosion will do less than 25% of its damage, you don't get hurt. 

That would result in minor nerfs to all the above except rockets and grellows.

**Comment by Largos** [comment]
"Yeah, it means aimed grenades are still just as vicious but unaimed / spammy / dodged grenades could be ignored rather than dealing a medium to small amount of damage."

Whichever option leads to this, I like.  So . . . a fixed option 2 from above?

I assume buildings are affected the same by all this?  Or do they undergo different explosion response properties?

**Comment by squeek502** [comment]
Was curious, so I made a table of how the radiuses would be affected if we scaled them based on where they currently do 0.5x damage or 0.25x damage.

| Weapon | Base Damage | Defined Radius | Damage at defined radius | Actual Radius | % of Base Damage At Actual Radius | 0.5x Damage Radius | Radius Difference (0.5 - Actual) | 0.25x Damage Radius | Difference (0.25 - Actual) |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| IC | 55 | 125 | -7.5 | 110 | 0.00% | 55 | -55.0 | 82.5 | -27.5 |
| Frag Gren | 145 | 270 | 10 | 270 | 6.90% | 145 | -125.0 | 217.5 | -52.5 |
| Railgun | 60 | 100 | 10 | 100 | 16.67% | 60 | -40.0 | 90 | -10.0 |
| Mirv | 180 | 270 | 45 | 270 | 25.00% | 180 | -90.0 | 270 | +0.0 |
| Mirvlet | 145 | 217.5 | 36.25 | 217.5 | 25.00% | 145 | -72.5 | 217.5 | +0.0 |
| Grellow Pipe | 108 | 130 | 43 | 130 | 39.81% | 108 | -22.0 | 162 | +32.0 |
| RPG | 102 | 100 | 52 | 100 | 50.98% | 102 | +2.0 | 153 | +53.0 |
| Blue Pipe | 108 | 100 | 58 | 100 | 53.70% | 108 | +8.0 | 162 | +62.0 |

Spreadsheet of the table: 
https://docs.google.com/spreadsheets/d/1YOUx3pYZSYi9P44bdAxJeMa-Jwmc7G-n0SWYjY3lQh0/edit?usp=sharing

**Comment by squeek502** [comment]
Did some more testing of TFC and was surprised by the results:
- https://github.com/fortressforever/tfc-reference-data/commit/2a8a9a52efa425ac587c62fffd23995a9a8df092
- https://github.com/fortressforever/tfc-reference-data/commit/69a64c7f86ce1fe9ada451f0df5aefd416d5a066

Here's a similar table showing radius falloff for TFC:

| Weapon | Base Damage | Radius | Damage at radius | % of Base Damage At Edge of Radius |
| --- | --- | --- | --- | --- |
| Frag Gren | 162 | 352 | 0 |  0.00% |
| Grellow Pipe | 108 | 120 | 45 | 41.67% |
| RPG | 83-101 (varies) | 109 max (varies) | ~40-50 (varies) | ~50% (very approximate) |

(MIRV and MIRVlets are the same as frag grenades)

---

So it seems like TFC also decouples radius, falloff, and damage, and it varies between damage sources, and that FF's damage falloff is more TFC-like than I thought.
